File: Mobile Applications/Labs/lab3_non_native/lib/view_model/reminder_list_view_model.dart
import 'package:flutter/material.dart';
import 'package:intl/intl.dart';

import '../model/reminder.dart';
import '../service/api_service.dart';
import '../utility/validator/reminder_validator.dart';

class ReminderListViewModel with ChangeNotifier {
  List<Reminder> reminders = [];
  final ApiService apiService = ApiService();
  final ReminderValidator reminderValidator = ReminderValidator();

  ReminderListViewModel() {
    loadReminders();
  }

  Future<void> loadReminders() async {
    reminders = await apiService.fetchReminders();
    notifyListeners();
  }

  void deleteReminder(int id) {
    reminders = reminders.where((reminder) => reminder.id != id).toList();
    apiService.saveReminders(reminders);
    notifyListeners();
  }

  void addReminder({
    required String title,
    required DateTime date,
    required DateTime time,
    required String category,
    required String priority,
    required String note,
  }) {
    final int highestId = reminders.fold(0, (int currentMaxId, Reminder reminder) {
          return currentMaxId > reminder.id ? currentMaxId : reminder.id;
        }) +
        1;

    final String dateString = DateFormat('dd/MM/yyyy').format(date);
    final String timeString = DateFormat('HH:mm').format(time);

    Reminder newReminder = Reminder(
      id: highestId,
      title: title,
      date: dateString,
      time: timeString,
      category: category,
      priority: priority,
      note: note,
      isCompleted: false,
    );

    try {
      reminderValidator.validate(newReminder);
    } catch (error) {
      rethrow;
    }

    reminders.add(newReminder);

    notifyListeners();

    apiService.saveReminders(reminders);
  }

  void updateReminder({
    required int id,
    required String title,
    required DateTime date,
    required DateTime time,
    required String category,
    required String priority,
    required String note,
  }) {
    String dateString = DateFormat('dd/MM/yyyy').format(date);
    String timeString = DateFormat('HH:mm').format(time);

    Reminder updatedReminder = Reminder(
      id: id,
      title: title,
      date: dateString,
      time: timeString,
      category: category,
      priority: priority,
      note: note,
    );

    try {
      reminderValidator.validate(updatedReminder);
    } catch (error) {
      rethrow;
    }

    int index = reminders.indexWhere((reminder) => reminder.id == id);
    if (index != -1) {
      reminders[index] = updatedReminder;
      notifyListeners();
    }

    apiService.saveReminders(reminders);
  }

  void updateReminderCompletion(Reminder reminder) {
    int index = reminders.indexWhere((r) => r.id == reminder.id);
    if (index != -1) {
      reminders[index] = reminder.updateCompletion();
      apiService.saveReminders(reminders);
      notifyListeners();
    }
  }
}

File: Mobile Applications/Labs/Lab2 (Native)/RemindMeNow/RemindMeNow/ViewModels/ReminderListViewModel.swift
import Foundation
import SwiftUI

class ReminderListViewModel: ObservableObject {
	
	@Published private var reminders: [Reminder] = []
	
	private let apiService: ApiService = ApiService()
	private let reminderValidator: ReminderValidator = ReminderValidator()
	
	init() {
		self.reminders = self.apiService.fetchReminders()
	}
	
	func getReminders() -> [Reminder] {
		return self.reminders
	}
	
	func deleteReminder(id: Int) {
		self.reminders = self.reminders.filter { $0.id != id }
		self.apiService.saveReminders(reminders: reminders)
	}
	
	func addReminder(
		title: String,
		date: Date,
		time: Date,
		category: String,
		priority: String,
		note: String
	) throws {
		let highestId = self.reminders.reduce(0) { (result, reminder) -> Int in
			return result > reminder.id ? result : reminder.id
		} + 1
		
		let dateTimeFormatter = DateFormatter()
		
		dateTimeFormatter.dateFormat = "dd/MM/yyyy"
		let dateString = dateTimeFormatter.string(from: date)
		
		dateTimeFormatter.dateFormat = "HH:mm"
		let timeString = dateTimeFormatter.string(from: time)
		
		let reminder = Reminder(id: highestId, title: title, date: dateString, time: timeString, category: category, priority: priority, note: note)
		
		do {
			try self.reminderValidator.validate(reminder: reminder)
		} catch {
			throw error
		}
		
		reminders.append(reminder)
		self.apiService.saveReminders(reminders: reminders)
	}
	
	func updateReminder(
		id: Int,
		title: String,
		date: Date,
		time: Date,
		category: String,
		priority: String,
		note: String
	) throws {
		let dateTimeFormatter = DateFormatter()
		
		dateTimeFormatter.dateFormat = "dd/MM/yyyy"
		let dateString = dateTimeFormatter.string(from: date)
		
		dateTimeFormatter.dateFormat = "HH:mm"
		let timeString = dateTimeFormatter.string(from: time)
		
		let reminder = Reminder(id: id, title: title, date: dateString, time: timeString, category: category, priority: priority, note: note)
		
		do {
			try self.reminderValidator.validate(reminder: reminder)
		} catch {
			throw error
		}
		
		if let index = reminders.firstIndex(where: { $0.id == id }) {
			reminders[index] = reminder
		}
		
		self.apiService.saveReminders(reminders: reminders)
	}

	
	func updateReminderCompletion(reminder: Reminder) {
		if let index = reminders.firstIndex(where: { $0.id == reminder.id }) {
			reminders[index] = reminder.updateCompletion()
		}
		
		self.apiService.saveReminders(reminders: reminders)
	}
}
